### PR TITLE
fix: prevent JSON parse error when loading data

### PR DESCRIPTION
**Pull request by eltigerchino**
fixes https://github.com/sveltejs/kit/issues/8751

This PR handles the edge case where loading the data of a non-existent prerendered dynamic route returns a 404 HTML page instead of JSON data. No test because our /basics test app has a root server layout load function, which causes it to do a native navigation instead, returning a 404 SSR response. The error only occurs when there is no root server layout load.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by dummdidumm** [comment]
Why is the response to the non-existent file returned as HTML and not as json? Would it be possible to solve this at the prerendering stage instead by creating a `__data.json` that tells the client "this is a 404" somehow?
Alternatively, if we detect that the return type is HTML, should we do a full page navigation instead? These cases are probably rare enough (and something's went wrong badly) that a full page reload is fine.

**Comment by eltigerchino** [comment]
updated to fallback to native navigation when receiving a HTML response instead of JSON when loading `__data.json`

File: packages/kit/src/runtime/client/client.js [hunk 1]
```diff
@@ -1889,50 +1891,53 @@ async function load_data(url, invalid) {
...
+						if (error) {
+							deferred.reject(deserialize(error));
+						} else {
+							deferred.fulfil(deserialize(data));
+						}
 					}
 				}
 			}
 		}
+		handle_stream();
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:1940]: What's this change about btw?

File: packages/kit/src/runtime/client/client.js [hunk 2]
```diff
@@ -1857,15 +1857,17 @@ async function load_data(url, invalid) {
 
 	const res = await native_fetch(data_url.href);
 
+	if (res.headers.get('content-type')?.includes('text/html')) {
+		await native_navigation(url);
```
Comment (dummdidumm) [inline packages/kit/src/runtime/client/client.js:1861]: I feel that's a better and more generic solution to our problem 👍 We should add a comment though explaining why this could happen.

File: packages/kit/src/runtime/client/client.js (hunk 1 above)
Comment (eltigerchino) [reply packages/kit/src/runtime/client/client.js:1940]: ah I was trying to resolve to the TODO about the eslint error https://eslint.org/docs/latest/rules/no-async-promise-executor I can revert it though

**Review by dummdidumm — APPROVED** [review]
